Alerting: Consume k8s Time Intervals API

tomratcliffe opened the pull request:
**What is this feature?**
(Conditionally) Consumes k8s-style APIs for managing mute timings for the Grafana alertmanager.

**Why do we need this feature?**
As part of the move to provide RBAC for managing alertmanager config entities, we need a more granular approach for the built in alertmanager.

This PR will use the kubernetes API if
* The alertmanager used is Grafana's
* The `alertingApiServer` feature toggle is enabled

**Who is this feature for?**

Alerting developers (for now). Will eventually help enable RBAC in more areas of the platform

tomratcliffe committed:
Make alerting API support passing in `body` as well as `data`

As the codegen APIs use `body` instead of the expected `data`, we need to map the properties so payloads are sent for mutations

tomratcliffe committed:
Remove transformResponse logic from generated time intervals API

As of now, we easily lose type information and it requires a lot of fiddling around. It's cleaner to just handle the processing within custom hooks

tomratcliffe committed:
Allow aligning table columns to end

tomratcliffe committed:
Add basic provisioning handlers

tomratcliffe committed:
Add basic time intervals k8s handlers

tomratcliffe committed:
Add handlers and method to allow configuring mock (non-grafana) alertmanager config

tomratcliffe committed:
Add tooltip for provisioning badge

tomratcliffe committed:
Use mock provisioning and time intervals handlers

tomratcliffe committed:
Update codegen timeintervals API

tomratcliffe committed:
Return dispatched action in delete mute timings so we can chain/await

tomratcliffe committed:
Add hooks for mute timings

tomratcliffe committed:
Create separate component for mute timing actions

tomratcliffe committed:
Use new hooks in mute timing form

tomratcliffe committed:
Refactor mute timings table to use new hooks and split out action buttons

tomratcliffe committed:
Refactor routing logic for mute timings page and use new hook

tomratcliffe committed:
Use new hook for showing number of mute timings

tomratcliffe committed:
Add additional tests for mute timings table and cover k8s logic

tomratcliffe committed:
Add util methods for k8s and refactor multi-interval display

tomratcliffe committed:
Remove mocks from mute timings tests and add additional checks

tomratcliffe committed:
Add metadata type to AM config

tomratcliffe committed:
Update translations, snapshots and betterer results

tomratcliffe committed:
Move export timings drawer to separate file

tomratcliffe committed:
Reuse return link for mute timings

tomratcliffe committed:
Add error case when mute timings cannot be loaded

tomratcliffe committed:
Use const for file provenance

tomratcliffe commented on public/app/features/alerting/unified/state/actions.ts:
```diff
@@ -608,31 +608,29 @@ export const deleteMuteTimingAction = (alertManagerSourceName: string, muteTimin
-                ...time_intervals_without_mute_to_save,
-              },
+    const { mute_time_intervals: _, ...configWithoutMuteTimings } = config?.alertmanager_config ?? {};
+    return withAppEvents(
```
Change here was just to return the dispatched action so we can consistently chain off it inside useMuteTimings

tomratcliffe committed:
Reuse const for file provenance

tomratcliffe commented on public/app/plugins/datasource/alertmanager/types.ts:
```diff
@@ -330,7 +331,8 @@ export interface TimeInterval {
   time_intervals: TimeInterval[];
-  provenance?: string;
+  provisioned?: boolean;
+  metadata?: ReadNamespacedTimeIntervalApiResponse['metadata'];
```
(I should possibly make a separate type that has this property on it instead of adding to this one here, as it's now directly from the datasource)

tomratcliffe committed:
Swap export & delete for more consistent experience with other Alerting pages

tomratcliffe commented:
/deploy-to-hg

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/consume-time-intervals-k8s-api

tomratcliffe commented:
/deploy-to-hg

tomratcliffe commented:
(I enabled the feature toggle on the ephemeral instance but there's an error from the API - will ask about it later today :+1:)

tomratcliffe committed:
Fix namespace logic

tomratcliffe commented:
/deploy-to-hg

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/consume-time-intervals-k8s-api

tomratcliffe commented:
/deploy-to-hg

tomratcliffe committed:
Fix incorrect return link when cancelling form

soniaAguilarPeiron commented on public/app/features/alerting/unified/MuteTimings.test.tsx:
```diff
@@ -143,169 +155,135 @@ const defaultConfigWithBothTimeIntervalsField: AlertManagerCortexConfig = {
 
-const resetMocks = () => {
-  jest.resetAllMocks();
+const expectToHaveRedirectedToTable = async () => expect(await screen.findByText(indexPageText)).toBeInTheDocument();
```
nit: I would change this variable name as its not a table Something like `expectToHaveRedirectedToRightRoute`

soniaAguilarPeiron commented on public/app/features/alerting/unified/MuteTimings.test.tsx:
```diff
@@ -143,169 +155,135 @@ const defaultConfigWithBothTimeIntervalsField: AlertManagerCortexConfig = {
-  jest.resetAllMocks();
+const expectToHaveRedirectedToTable = async () => expect(await screen.findByText(indexPageText)).toBeInTheDocument();
+
+const fillOutForm = async ({
```
Nice!❤️

soniaAguilarPeiron commented on public/app/features/alerting/unified/NotificationPolicies.tsx:
```diff
@@ -65,6 +65,7 @@ const AmRoutes = () => {
 
   const { selectedAlertmanager, hasConfigurationAPI, isGrafanaAlertmanager } = useAlertmanager();
   const { getRouteGroupsMap } = useRouteGroupsMatcher();
+  const { data: muteTimings } = useMuteTimings({ alertmanager: selectedAlertmanager ?? '' });
```
maybe we should default to `GRAFANA_RULES_SOURCE_NAME` instead of an empty string?

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/DynamicTable.tsx:
```diff
@@ -108,7 +110,7 @@ export const DynamicTable = <T extends object>({
           {isExpandable && <div className={styles.cell} />}
           {cols.map((col) => (
-            <div className={styles.cell} key={col.id}>
+            <div className={styles.cell} key={col.id} style={{ justifyContent: col.alignColumn || 'initial' }}>
```
why are we using style here? instead of including this style in the classname?

soniaAguilarPeiron commented on public/app/features/alerting/unified/components/mute-timings/useExportMuteTimingsDrawer.tsx:
```diff
@@ -0,0 +1,39 @@
+    toggleShowExportDrawer(false);
+  }, [toggleShowExportDrawer]);
+
+  const handleOpen = (receiverName: string | typeof ALL_MUTE_TIMINGS) => {
```
`receivername` should be `muteTimingName` I guess?

soniaAguilarPeiron reviewed (approved):
LGTM! 🚀 I added a non blocking comments

gillesdemey commented on public/app/features/alerting/unified/MuteTimings.tsx:
```diff
@@ -1,80 +1,53 @@
+const EditTimingRoute = () => {
   const [queryParams] = useQueryParams();
   const { selectedAlertmanager } = useAlertmanager();
-  const { currentData, isLoading, error } = useAlertmanagerConfig(selectedAlertmanager, {
```
Agree with separate refactor, though I don't believe many users will have so many mute timings that this might become an issue any time soon.

tomratcliffe replied:
Yep, a refactor is more just for a cleaner separation of the logic in our routes etc :+1:

gillesdemey commented on public/app/features/alerting/unified/NotificationPolicies.tsx:
```diff
@@ -65,6 +65,7 @@ const AmRoutes = () => {
 
   const { selectedAlertmanager, hasConfigurationAPI, isGrafanaAlertmanager } = useAlertmanager();
   const { getRouteGroupsMap } = useRouteGroupsMatcher();
+  const { data: muteTimings } = useMuteTimings({ alertmanager: selectedAlertmanager ?? '' });
```
Is adding support for pagination a prerequisite to get the totals from the API?

tomratcliffe replied:
Not sure actually - but right now, as the limit doesn't do anything, I have to query for everything anyway and check the length. If `limit` was functional, then I could ask for 1 item and check the metadata - but I don't think it'll matter too much for now?

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/consume-time-intervals-k8s-api

tomratcliffe committed:
Change dynamic table col alignment to reuse styles

tomratcliffe commented on public/app/features/alerting/unified/NotificationPolicies.tsx:
```diff
@@ -207,9 +208,8 @@ const AmRoutes = () => {
-  const time_intervals = result?.alertmanager_config ? mergeTimeIntervals(result?.alertmanager_config) : [];
 
-  const numberOfMuteTimings = time_intervals.length;
+  const numberOfMuteTimings = muteTimings?.length || 0;
```
I agree - however, the reason that this one is potentially undefined is because it's driven by RTKQ-style hooks, and `muteTimings` is potentially undefined if the API request hasn't yet completed. What this should/could do is show a loading spinner for the count badge while its loading. That could come along with a refactor to refactor the routes so this logic isn't even in here?

I considered returning a loading component if the timings were still fetching, but that would mean the entire policies page would be hidden until a small part of the UI had completed loading

tomratcliffe committed:
Reuse form loading state

tomratcliffe committed:
Update provisioning tooltip to include provisioning type

tomratcliffe committed:
Use provisioning data in mute timings table tooltip

tomratcliffe committed:
Fix arg name from copy/paste

tomratcliffe committed:
Rename helper method

tomratcliffe committed:
Update provisioning check logic

tomratcliffe committed:
Update types and aliases for time intervals

tomratcliffe committed:
Update translations

tomratcliffe committed:
Tidy up mute timing types

tomratcliffe committed:
Use autogenerated RTKQ hooks

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/consume-time-intervals-k8s-api

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/consume-time-intervals-k8s-api

tomratcliffe committed:
Revert i18n changes

konrad147 commented on public/app/plugins/datasource/alertmanager/types.ts:
```diff
@@ -330,7 +330,7 @@ export interface TimeInterval {
 export type MuteTimeInterval = {
   name: string;
   time_intervals: TimeInterval[];
-  provenance?: string;
```
Why the change? Isn't it the API response type?

gillesdemey replied:
I _think_ this is the "UI" data model and not the DTO.

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/consume-time-intervals-k8s-api

tomratcliffe committed:
Merge remote-tracking branch 'origin/main' into alerting/consume-time-intervals-k8s-api

tomratcliffe committed:
PR review comments

tomratcliffe commented:
@gillesdemey do you have any thoughts on this? And @konrad147 I've addressed your comments, what do you think? Are we able to merge this in (as its also behind a feature toggle still)?

tomratcliffe commented on public/app/features/alerting/unified/components/mute-timings/useMuteTimings.tsx:
```diff
@@ -0,0 +1,343 @@
+      getAlertmanagerTimeIntervals(alertmanager);
+    }
+  }, [alertmanager, getAlertmanagerTimeIntervals, getGrafanaTimeIntervals, useK8sApi]);
+  return useK8sApi ? intervalsResponse : configApiResponse;
```
(as per our earlier chat, this does in fact include the loading/error/all other RTKQ properties, as its the whole response object)

konrad147 reviewed (approved):
LGTM!

gillesdemey reviewed (approved):
just a few nits; feel free to ignore those! Great work! 🚀

tomratcliffe committed:
Remove coercion for updating time intervals
